exercises/static/exercises/montecarlo_laser_loc/python_template/ros2_humble/HAL.py

```diff
@@ -1,48 +1,67 @@
 import rclpy
 import threading
+import time
 
-from interfaces.motors import PublisherMotors
-from interfaces.pose3d import ListenerPose3d
-from interfaces.laser import ListenerLaser
-from interfaces.bumper import ListenerBumper
-
+from hal_interfaces.general.motors import MotorsNode
+from hal_interfaces.general.odometry import OdometryNode
+from hal_interfaces.general.laser import LaserNode
+from hal_interfaces.general.bumper import BumperNode
 
 # Hardware Abstraction Layer
 IMG_WIDTH = 320
 IMG_HEIGHT = 240
 
-rclpy.create_node('HAL')
-
-motors = PublisherMotors("/cmd_vel", 4, 0.3)
-pose3d = ListenerPose3d("/odom")
-laser = ListenerLaser("/roombaROS/laser/scan")
-bumper = ListenerBumper("/roombaROS/events/bumper","roombaROS")
+freq = 30.0
 
-# Spin nodes so that subscription callbacks load topic data
-# Bumper has to be spinned differently so that GetEntityState works
-executor = rclpy.executors.MultiThreadedExecutor()
-executor.add_node(pose3d)
-executor.add_node(laser)
-executor_thread = threading.Thread(target=executor.spin, daemon=True)
-executor_thread.start()
+print("HAL initializing", flush=True)
+if not rclpy.ok():
+    rclpy.init(args=None)
 
-print("HAL-Nodes Thread Started")
+    ### HAL INIT ###
+    motor_node = MotorsNode("/cmd_vel", 4, 0.3)
+    odometry_node = OdometryNode("/odom")
+    laser_node = LaserNode("/roombaROS/laser/scan")
+    bumper_node = BumperNode(
+        [
+            "/roombaROS/events/right_bumper",
+            "/roombaROS/events/center_bumper",
+            "/roombaROS/events/left_bumper",
+        ]
+    )
 
-def getBumperData():
-    bumper.spin_bumper_node()
-    return bumper.getBumperData()
+    # Spin nodes so that subscription callbacks load topic data
+    executor = rclpy.executors.MultiThreadedExecutor()
+    executor.add_node(odometry_node)
+    executor.add_node(laser_node)
```

Comment:
Hi @javizqh Why is bumper node not added in the executor?

Reply:
As the bumper is not as essential as the other nodes and I think it is not used a lot in the solution that is why I kept the spinOnce method

Reply:
But if you want I could add it to the executor

Reply:
oh okay, that makes sense...